### DISCUSSION TITLE
Passing variables into client.query

### DISCUSSION THREAD

**Discussion by mooyg** [urql-graphql/urql#2040] [Q&A] [answered]
```
interface UseLazyQueryArgs {
  query: TypedDocumentNode<any, object>
  variables?: Record<string, unknown>
}

type UseLazyQueryResponse = [OperationResult<any, object> | undefined, () => void]
export const useLazyQuery = ({ query, variables }: UseLazyQueryArgs): UseLazyQueryResponse => {
  const client = useClient()
  const accessToken = useAccessToken()
  const [data, setData] = useState<OperationResult<any, object>>()
  console.log(variables)
  const execQuery = client.query(query, variables, {
    fetchOptions: () => {
      return {
        headers: { Authorization: `Bearer ${accessToken}` },
      }
    },
  })

  return [
    data,
    () => {
      execQuery.toPromise().then((result) => setData(result))
    },
  ]
}
``` 
trying to pass in variables 
```
  const [foo, getPages] = useLazyQuery({
    query: GET_PAGES_BY_TEMPLATEID,
    variables: {
      templateId: templateId,
    },
  })
``` 
when running the function `getPages()` it returns `"[GraphQL] Variable "$TemplateId" of required type "String!" was not provided."` as the error 
query looks like 
```
export const GET_PAGES_BY_TEMPLATEID = gql`
  query GetPagesByTemplateId($TemplateId: String!) {
    getPagesByTemplateId(templateId: $TemplateId) {
      name
      id
      templateId
    }
  }
```

**Comment by JoviDeCroock** [comment, marked as answer]
In your query you wrote `TemplateId` notice the capital T and in your `variables` you wrote `templateId` with lower case t

**Reply by mooyg** [reply]
> In your query you wrote `TemplateId` notice the capital T and in your `variables` you wrote `templateId` with lower case t

Ah my bad. `"[GraphQL] Variable "$templateId" of non-null type "String!" must not be null."` idk why it is considering it as `null`?

**Reply by JoviDeCroock** [reply]
You can log it out 😅 maybe you're passing in null

**Reply by mooyg** [reply]
> You can log it out 😅 maybe you're passing in null

![image](https://user-images.githubusercontent.com/61348580/138457363-32e250ab-e309-4a2a-b81c-51c6ef8ccce1.png)
seems fine

**Reply by JoviDeCroock** [reply]
I mean, I think it's because you're executing the function in advance, I think this is what you're looking for:

```js
const useLazyQuery = ({ query, variables }) => {
  const client = useClient()
  const [data, setData] = useState()

  return [
    data,
    () => {
      client.query(query, variables).toPromise().then((result) => setData(result))
    },
  ]
}
```

Example: https://codesandbox.io/s/zen-fog-nbcdy?file=/src/App.js

**Reply by mooyg** [reply]
> variables

Hmm 

> I mean, I think it's because you're executing the function in advance, I think this is what you're looking for:
> 
> ```
> const useLazyQuery = ({ query, variables }) => {
>   const client = useClient()
>   const [data, setData] = useState()
> 
>   return [
>     data,
>     () => {
>       client.query(query, variables).toPromise().then((result) => setData(result))
>     },
>   ]
> }
> ```

hmm but I am passing variables too so idk what's the difference. Tried this and it seems to reproduce same error. Checked network tab 
![image](https://user-images.githubusercontent.com/61348580/138458501-088de9a9-67e7-45b8-b01f-5d134ba17663.png)

**Reply by JoviDeCroock** [reply]
I mean if it's in the network-tab like that it means that everything is passed correctly from our side, then this server is only coming from your server.

Made a small example without the server throwing errors 😅 https://codesandbox.io/s/zen-fog-nbcdy?file=/src/App.js you should look at your schema whether it's expected or needs to be wrapped in an input, .... to be passed in

**Reply by mooyg** [reply]
> I mean if it's in the network-tab like that it means that everything is passed correctly from our side, then this server is only coming from your server.

hmm but seems to be working fine in playground 
![image](https://user-images.githubusercontent.com/61348580/138459188-196c9947-12e4-4e55-a80c-1460dd37dee5.png)

**Reply by JoviDeCroock** [reply]
You're not doing the same thing as your query though, you are filling in the variable here instead of using the variables tab of your playground

**Reply by mooyg** [reply]
> You're not doing the same thing as your query though, you are filling in the variable here instead of using the variables tab of your playground

![image](https://user-images.githubusercontent.com/61348580/138459945-be10831c-22c5-4cf8-a936-60fae032d5aa.png)
also `templateId` and stuff logs correct even when making request from the client instead of playground

(sorry didn't notice the example checking it)

Edit: Hmmm so what I am guessing is it makes initial request with `null` `templateId` cause it is null initially then when I actually call the function with the variable it just gets it from the cache ? 

Edit: Seems to be the case 👍🏻
